### DISCUSSION TITLE
PowerOnState during warm restart on Tasmota 13.2.0.3 ESP32-C3 (Shelly Plus 1 Mini device)

### DISCUSSION THREAD

**Discussion by phidix** [arendst/Tasmota#20144] [Q&A] [answered]
Hello,
Yesterday I've migrated **Shelly Plus 1 Mini** device to **Tasmota 13.2.0.3 ESP32-C3**. Migration process went flawlessly, but I can't solve the issue with **PowerOnState0** during **warm restarts**.
I've invoked the command: `PowerOnState0 0` in order to have relay automatically set to `OFF` after **initial boots (hard restarts of device)** but for some reason this setting affects also every **warm restart** setting **relay** to **Off**.
Do you know whether this is a bug or expected behavior for this device? I'm using **Tasmota v.11** for **Sonoff Mini device** and I don't observe such behavior - after **warm restarts** state of the relay doesn't change.

Below I attach log files gathered just after "warm restart":

```
00:00:00.001 HDW: ESP32-C3 v0.4 
00:00:00.012 UFS: FlashFS mounted with 340 kB free
00:00:00.018 CFG: Loaded from File, Count 104
00:00:00.023 QPC: Count 1
00:00:00.027 SCR: ufilesystem found
00:00:00.077 BRY: Berry initialized, RAM used 4976 bytes
00:00:00.099 Project tasmota - Tasmota Version 13.2.0.3(tasmota32)-2_0_14(2023-12-02T00:47:55)
00:00:00.148 TAP: Loaded Tasmota App 'Partition_Wizard.tapp'
00:00:00.894 WIF: Connecting to AP1 *** Channel 7 BSSId *** in mode 11n as tasmota-***...
00:00:01.720 WIF: Connected
00:00:01.981 HTP: Web server active on tasmota-*** with IP address 192.***
15:27:31.794 RSL: INFO1 = {"Info1":{"Module":"Shelly Plus1Mini","Version":"13.2.0.3(tasmota32)","FallbackTopic":"cmnd/DVES_98B920_fb/","GroupTopic":"cmnd/tasmotas/"}}
15:27:31.806 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-***","IPAddress":"192.***","IP6Global":"","IP6Local":"fe***20%st1"}}
15:27:31.819 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset digital core","BootCount":30}}
15:27:31.831 RSL: RESULT = {"POWER":"OFF"}
15:27:31.832 RSL: POWER = OFF
15:27:34.966 QPC: Reset
15:27:34.970 RSL: STATE = {"Time":"2023-12-02T15:27:34","Uptime":"0T00:00:06","UptimeSec":6,"Heap":154,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":24,"Objects":303},"POWER":"ON","Wifi":{"AP":1,"SSId":"b***","BSSId":"***","Channel":7,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:02"}}
15:27:35.006 RSL: SENSOR = {"Time":"2023-12-02T15:27:35","Switch1":"OFF","ANALOG":{"Temperature1":67.8},"TempUnit":"C"}
```

**Comment by barbudor** [comment]
I see the same behavior
unexpected
I'm investingating 🕵️

**Reply by barbudor** [reply]
Same behavior on ESP32C3 and ESP32 using either v11.0 or v13.0.3.
The difference is not related to Tasmota version but to a difference between ESP8266 and ESP32C3/ESP32
Still it seems that it should be possible to distinguish between restart and reset/power-cycle:

esp8266 : 
restart     : "RestartReason":"Software/System restart"
reset       : "RestartReason":"External System"
power cycle : "RestartReason":"External System"

esp32c3 :
restart     : "RestartReason":"Software reset digital core"
reset       : "RestartReason":"Vbat power on reset"
power cycle : "RestartReason":"Vbat power on reset"

esp32 :
restart     : "RestartReason":"Software reset CPU"
reset       : "RestartReason":"Vbat power on reset"
power cycle : "RestartReason":"Vbat power on reset"

looking further...

**Comment by arendst** [comment]
Likely caused by ``GpioForceHoldRelay()`` keeping the cripple ESP32C3 from toggling the relay at restart.

I'll have a look at ``PowerOnState0 0`` impact.....

**Reply by barbudor** [reply]
I would look into `bool ResetReasonPowerOn(void)`

The function is not CPU dependant and is based on REASON_DEFAULT_RST (0) or REASON_EXT_SYS_RST (6)

But the value I see on ESP32/C3 is POWERON_RESET          =  1 "Vbat power on reset"

**Reply by arendst** [reply]
Yeah, as you seem to notice it doesn't work on ESP32 too that's more obvious.

To handel the ESP32 restart codes like the ESP8266 I had to add a comparison table. See ``ESP_ResetInfoReason()``. For some reason I decided to have the ESP32 "Software reset digital core" compare to the ESP8266 "REASON_EXT_SYS_RST"...

**Reply by barbudor** [reply]
Mistake above: the value on cold restart on ESP32 is POWERON_RESET          =  1 "Vbat power on reset"

**Reply by barbudor** [reply]
May be it would be needed to check on every variant of ESP32. I only have ESP32 and C3

**Comment by arendst** [comment, marked as answer]
I think best solution is making below change in file ``support_esp32.ino``:
```
uint32_t ESP_ResetInfoReason(void) {
  RESET_REASON reason = rtc_get_reset_reason(0);
  if (1  == reason) { return REASON_DEFAULT_RST; }       // POWERON_RESET
  if ((3  == reason) || (12 == reason)) { return REASON_SOFT_RESTART; }  // SW_RESET / RTC_SW_SYS_RESET and SW_CPU_RESET / RTC_SW_CPU_RESET
  if (5  == reason) { return REASON_DEEP_SLEEP_AWAKE; }  // DEEPSLEEP_RESET
//  if (3  == reason) { return REASON_EXT_SYS_RST; }       // SW_RESET / RTC_SW_SYS_RESET
  return -1; //no "official error code", but should work with the current code base
}
```

**Reply by phidix** [reply]
Thank you!
I've checked this change with my device  **Shelly Plus 1 Mini device** **ESP32-C3** and it resolves the issue for it.

**Comment by arendst** [comment]
Try latest dev branch with above change included.

**Reply by barbudor** [reply]
Same

**Reply by arendst** [reply]
Strange. It worked for me on my athom esp32c3 power plug.

Setting PowerOnState to 0 controls the relay on powerplug ac cycle. It doesn't control the relay on a restart. Both as expected.

**Reply by barbudor** [reply]
trying again

**Reply by barbudor** [reply]
Just testing on an ESP32 devkit : 
```
18:19:04.534 CMD: power on
18:19:04.545 MQT: stat/esp32dev3/POWER = {"POWER":"ON"}
18:19:04.549 MQT: stat/esp32dev3/POWER = ON (retained)
18:19:09.046 CMD: poweron state
18:19:09.057 MQT: stat/esp32dev3/COMMAND = {"Command":"Unknown"}
18:19:15.401 CMD: poweronstate
18:19:15.408 MQT: stat/esp32dev3/POWERONSTATE = {"PowerOnState":0}
18:19:17.955 CMD: power on
18:19:17.966 MQT: stat/esp32dev3/POWER = {"POWER":"ON"}
18:19:17.970 MQT: stat/esp32dev3/POWER = ON (retained)
18:19:22.367 CMD: restart 1
18:19:22.375 MQT: stat/esp32dev3/RESTART = {"Restart":"Restarting"}
....
18:19:35.945 MQT: tele/esp32dev3/INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":13}}
18:19:35.962 MQT: stat/esp32dev3/POWER = {"POWER":"OFF"}
```

Added also so serial print at crucial points:
```
00:00:00.097 QPC: Count 1
ESP_ResetInfoReason():reason=12
ESP_ResetInfoReason():reason=12
ResetReasonPowerOn():reset_reason=4
ResetReasonPowerOn():is_power_on=0
00:00:00.196 BRY: Berry initialized, RAM used 3711 bytes
ESP_ResetInfoReason():reason=12
ESP_ResetInfoReason():reason=12
ResetReasonPowerOn():reset_reason=4
ResetReasonPowerOn():is_power_on=0
00:00:00.213 Project tasmota - esp32dev3 Version 13.2.0.3(bhome)-2_0_14(2023-12-02T18:16:46)
00:00:06.930 QPC: Reset
00:00:07.778 WIF: Connecting to AP1 BNETWLAN Channel 1 BSSId 88:C3:97:F2:25:EC in mode 11n as esp32dev3...
00:00:09.760 WIF: Connected
00:00:09.916 HTP: Web server active on esp32dev3 with IP address 192.168.168.196
18:19:35.877 MQT: Attempting connection...
18:19:35.903 MQT: Connected
18:19:35.908 MQT: tele/esp32dev3/LWT = Online (retained)
18:19:35.910 MQT: cmnd/esp32dev3/POWER = 
ESP_ResetInfoReason():reason=12
```

So correctly identified as a warm restart but still switch power off
And before you ask, yes savedate is on

**Reply by barbudor** [reply]
exact same logs with esp32 C3 dev board

**Reply by arendst** [reply]
Will test again tomorrow..

**Reply by barbudor** [reply]
Works for me too
SO0 was off
